rig/machine_control/machine_controller.py

```diff
@@ -1591,6 +1671,80 @@ def seek(self, n_bytes, from_what=os.SEEK_SET):
             )
 
 
+class _WriteBuffer(object):
+    """Write buffer used by :py:class:`.MemoryIO` to combine multiple writes
+    together.
+    """
+
+    def __init__(self, x, y, p, controller, buffer_size=0):
+        self.x = x
+        self.y = y
+        self.p = p
+        self.controller = controller
+
+        # A buffer of writes
+        self.buffer = bytearray(buffer_size)
+        self.start_address = None
+
+        self.current_end = 0
+        self.buffer_size = buffer_size
+
+    def add_new_write(self, start_address, data):
+        """Add a new write to the buffer."""
+        if len(data) > self.buffer_size:
+            # Perform the write if we couldn't buffer it at all
+            self.flush()  # Flush to ensure ordering is preserved
+            self.controller.write(start_address, data,
+                                  self.x, self.y, self.p)
+            return
+
+        if self.start_address is None:
+            # No value currently buffered, add this one
+            self.start_address = start_address
+
+        # If we can fit this write into the buffer then include it,
+        # otherwise we flush the current buffer and start again
+        start_offset = start_address - self.start_address
+        end_offset = start_offset + len(data)  # Byte AFTER the end of the data
+
+        if start_offset <= self.current_end and end_offset <= self.buffer_size:
+            # The write is entirely contained within the buffer and starts
+            # within the area of the buffer which already contains data, so we
+            # can just modify the buffer.
+            self.buffer[start_offset:end_offset] = data
+            self.current_end = max(end_offset, self.current_end)
```

Comment:
I think the fact that not every write made will actually get to the hardware (when buffering is enabled) probably warrants documenting, though I can't really see it ever biting someone... if you care about exactly what writes are happening you should be flushing accordingly...